[Grid] Allow extrinsically sized grid to perform simplified layout when grid items are mutated.

sammygill opened the pull request:
#### 83bba6965f882753ecfbea7c0e45ad3b6ea3dd50
<pre>
[Grid] Allow extrinsically sized grid to perform simplified layout when grid items are mutated.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290573">https://bugs.webkit.org/show_bug.cgi?id=290573</a>
<a href="https://rdar.apple.com/148055966">rdar://148055966</a>

Reviewed by Alan Baradlay.

When a grid item experiences some types of mutations, such as content
mutations or non-grid specific style changes, it may be possible to
contain the damage within the formatting context established by the item
and not need to propagate this damage into the formatting context of the
grid itself or any ancestors. One such example of content where this is
a case is when a grid item experiences a content mutation but is inside
a grid that is extrinsically sized by perhaps having all fixed or
percentage-based rows and columns. In this case, the damage is local to
the grid item and will have no impact on the GFC and will not change any
of the constraints on the grid item.

This patch attempts to identify these types of damage and respond by
running simplified layout from the RenderView, or any other layout root,
through the grid. We can do this as we are invalidating the containing
block chain and seeing if we find an ancestor that meets this criterion.
If so, we will start marking needs simplified layout on the chain.

The mechanism that is used is a new function on RenderGrid, which
is not meant to be completely exhaustive but focuses on some
simple pieces of content and can be built upon in future patches.

While most of the conditions we are looking at currently are fairly
straightforward, I should probably mention and explain my rationale for
a few notable ones:

1.) We check to see if the grid has a logical width of auto but couple this
with that fact that the grid needs to be participating in block layout/a
BFC. In some formatting contexts, having a logical width of auto may end
up performing some sort of measurements on the box based on the content,
but in this case, it should not, so we can allow it. If we need to check
whether or not the grid is extrinsically sized when it lives inside
another type of formatting context, we may need to revisit this logic to
make it correct.

2.) Checking to see if the grid has an aspect ratio. This is to avoid
the automatic content-based minimum size that may be applied by the use
of the property.

3.) Making sure all of the items are placed within extrinsically sized
explicit (i.e. grid-template-rows/columns) tracks. The purpose here is
twofold. First, if any items need item placement, which could happen by
adding or removing grid items, we should just bail out since at that
point the GFC itself experienced damage and we cannot predict what will
happen. Second, in a previous layout, a grid item may have been placed in
a track that was implicitly created. If that is the case, we need to
figure out how that track was sized since intrinsic values could have
been used for grid-auto-rows/columns, but we bail now for simplicity.

In order to accomplish the above, we needed to shuffle around the logic
in RenderTreeBuilder::attachToRenderGrid so that we mark the grid as
needing item placement before we attach the new renderer to the grid.
This is because we can only determine if all items are placed within
Extrinsic tracks if we do not need to perform item placement. Otherwise,
we could add a renderer to the grid, start marking the containing block
chain, and incorrectly determine we can mark the grid as needing
simplified layout before we mark it as needing item placement.

I also added various tests in grid-extrinsically-sized-mutations.html
to help test that we were still performing layout properly. The tests
include changes to the grid items only along with mutations on the grid
and the constraints imposed upon it by its containing block.

* LayoutTests/imported/w3c/web-platform-tests/css
* LayoutTests/imported/w3c/web-platform-tests/css
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::isExtrinsicallySized const)
(WebCore::RenderGrid::isPlacedWithinExtrinsically
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::markContainingBlocksForLa
* Source/WebCore/rendering/updating/RenderTreeBui
(WebCore::RenderTreeBuilder::attachToRenderGrid):

Canonical link: <a href="https://commits.webkit.org/292994@main">https://commits.webkit.org/292994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be04a747fc549542a35c93ba278d4317c71a7864

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97540 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17165 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102705 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48069 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99585 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25618 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74359 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31506 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100543 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54705 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6092 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47511 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83022 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6171 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104647 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24620 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17961 "Found 3 new test failures: fast/workers/use-machine-stack.html imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-pause-networkState.html imported/w3c/web-platform-tests/workers/Worker-timeout-cancel-order.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83373 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24992 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84334 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82794 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20868 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27337 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5028 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18213 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24581 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29750 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24403 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27717 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25977 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBlock.h:
```diff
@@ -122,6 +122,8 @@ class RenderBlock : public RenderBox {
     // FIXME-BLOCKFLOW: Remove virtualizaion when all of the line layout code has been moved out of RenderBlock
     virtual bool containsFloats() const { return false; }
 
+    virtual bool isExtrinsicallySized() const { return false; }
```
Let's keep it on RenderGrid until we implement a non-grid variant.

alanbaradlay commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -65,6 +65,52 @@ RenderGrid::RenderGrid(Element& element, RenderStyle&& style)
+        if (isInline())
+            return false;
+        auto* containingBlock = this->containingBlock();
+        return containingBlock && containingBlock->isBlockContainer();
```
This is OK but the spec accurate check would be looking at the containing block (block container) and see if it has inline or block level boxes (`containingBlock->childrenInline()`)

sammygill replied:
Good point!

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -677,6 +678,9 @@ RenderElement* RenderObject::markContainingBlocksForLayout(RenderElement* layout
                 simplifiedNormalFlowLayout = true;
         }
 
+        if (auto* renderBlock = dynamicDowncast<RenderBlock>(container.get()); renderBlock && renderBlock->isExtrinsicallySized())
```
should be limited to RenderGrid for now.

sammygill committed:
[Grid] Allow extrinsically sized grid to perform simplified layout when grid items are mutated.
https://bugs.webkit.org/show_bug.cgi?id=290573
rdar://148055966

Reviewed by Alan Baradlay.

When a grid item experiences some types of mutations, such as content
mutations or non-grid specific style changes, it may be possible to
contain the damage within the formatting context established by the item
and not need to propagate this damage into the formatting context of the
grid itself or any ancestors. One such example of content where this is
a case is when a grid item experiences a content mutation but is inside
a grid that is extrinsically sized by perhaps having all fixed or
percentage-based rows and columns. In this case, the damage is local to
the grid item and will have no impact on the GFC and will not change any
of the constraints on the grid item.

This patch attempts to identify these types of damage and respond by
running simplified layout from the RenderView, or any other layout root,
through the grid. We can do this as we are invalidating the containing
block chain and seeing if we find an ancestor that meets this criterion.
If so, we will start marking needs simplified layout on the chain.

The mechanism that is used is a new function on RenderGrid, which
is not meant to be completely exhaustive but focuses on some
simple pieces of content and can be built upon in future patches.

While most of the conditions we are looking at currently are fairly
straightforward, I should probably mention and explain my rationale for
a few notable ones:

1.) We check to see if the grid has a logical width of auto but couple this
with that fact that the grid needs to be participating in block layout/a
BFC. In some formatting contexts, having a logical width of auto may end
up performing some sort of measurements on the box based on the content,
but in this case, it should not, so we can allow it. If we need to check
whether or not the grid is extrinsically sized when it lives inside
another type of formatting context, we may need to revisit this logic to
make it correct.

2.) Checking to see if the grid has an aspect ratio. This is to avoid
the automatic content-based minimum size that may be applied by the use
of the property.

3.) Making sure all of the items are placed within extrinsically sized
explicit (i.e. grid-template-rows/columns) tracks. The purpose here is
twofold. First, if any items need item placement, which could happen by
adding or removing grid items, we should just bail out since at that
point the GFC itself experienced damage and we cannot predict what will
happen. Second, in a previous layout, a grid item may have been placed in
a track that was implicitly created. If that is the case, we need to
figure out how that track was sized since intrinsic values could have
been used for grid-auto-rows/columns, but we bail now for simplicity.

In order to accomplish the above, we needed to shuffle around the logic
in RenderTreeBuilder::attachToRenderGrid so that we mark the grid as
needing item placement before we attach the new renderer to the grid.
This is because we can only determine if all items are placed within
Extrinsic tracks if we do not need to perform item placement. Otherwise,
we could add a renderer to the grid, start marking the containing block
chain, and incorrectly determine we can mark the grid as needing
simplified layout before we mark it as needing item placement.

I also added various tests in grid-extrinsically-sized-mutations.html
to help test that we were still performing layout properly. The tests
include changes to the grid items only along with mutations on the grid
and the constraints imposed upon it by its containing block.

* LayoutTests/imported/w3c/web-platform-tests/css
* LayoutTests/imported/w3c/web-platform-tests/css
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::isExtrinsicallySized const)
(WebCore::RenderGrid::isPlacedWithinExtrinsically
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::markContainingBlocksForLa
* Source/WebCore/rendering/updating/RenderTreeBui
(WebCore::RenderTreeBuilder::attachToRenderGrid):

Canonical link: https://commits.webkit.org/292994@main
